### PR TITLE
Dynamic provisioner for AWS EBS

### PR DESCRIPTION
**Pull request by pmorie**
Replaces #15603. 

@thockin @saad-ali @jsafrane @markturansky

### CODE REVIEW

**Comment by saad-ali** [comment]
Whoops I was looking at https://github.com/kubernetes/kubernetes/pull/15603. Looking at this now

File: pkg/volume/aws_ebs/aws_util.go [hunk 1]
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
...
+func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (string, int, error) {
+	const mega = 1024 * 1024
+	cloud := c.plugin.host.GetCloudProvider()
+	if cloud == nil {
+		glog.Errorf("Cloud provider not initialized properly")
+		return "", 0, errors.New("Cloud provider not initialized properly")
+	}
+
+	// convert to MB with rounding
+	volSize := int((c.options.Capacity.Value() + mega - 1) / mega)
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_util.go]: Could you add a comment on how this work. Are the units assumed?

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 2]
```diff
@@ -349,3 +390,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
...
+var _ volume.Provisioner = &awsElasticBlockStoreProvisioner{}
+
+func (c *awsElasticBlockStoreProvisioner) Provision(pv *api.PersistentVolume) error {
+	volumeID, size, err := c.manager.CreateVolume(c)
+	if err != nil {
+		return err
+	}
+	pv.Spec.PersistentVolumeSource.AWSElasticBlockStore.VolumeID = volumeID
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%d", size)),
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_ebs.go]: No units suffix? Are they assumed?

File: pkg/volume/aws_ebs/aws_util.go [hunk 3]
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
 	}
 	return nil
 }
+
+func (util *AWSDiskUtil) DeleteVolume(d *awsElasticBlockStoreDeleter) error {
+	cloud := d.plugin.host.GetCloudProvider()
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_util.go]: nit: this code to fetch and cast the cloud provider gets repeated a lot, factor it out into a util method, if possible.

**Comment by saad-ali** [comment]
A few small comments, but otherwise looks good.

**Comment by markturansky** [comment]
@pmorie @saad-ali I suppose the question now is whether we go with these new ones or revert back to @jsafrane's PRs.  We thought we'd be able to get these merged quickly after 14537 and still hit our deadline, but that has already slipped.  I'm thinking we should revert to the original author who owns the feature, since Jan will also be the one improving these same provisioners for many volume types in the next iteration of provisioning.  He'll have all these in shape by Monday morning (he's in Europe and he'll have these rebased before the rest of us get to work).

Hat tip to Paul for stepping up and trying to help get the ball over the goal line.  It's much appreciated.

File: pkg/volume/aws_ebs/aws_util.go (hunk 3 above)
Comment (jsafrane) [inline pkg/volume/aws_ebs/aws_util.go]: Good point, fixed

**Comment by pmorie** [comment]
@markturansky @saad-ali I say we just keep these open.  Jan is an owner in my repo now so he can update these branches with feedback without involving me.

**Comment by markturansky** [comment]
Yeah, Jan's already doing that and closed the originals.

**Comment by jsafrane** [comment]
Yes, @pmorie, please squash whatever is needed if we get LGTM when I sleep... See you tomorrow.

**Comment by pmorie** [comment]
@jsafrane will do

**Comment by saad-ali** [comment]
Cool, I'll take a look at this now.

**Comment by saad-ali** [comment]
LGTM

**Comment by saad-ali** [comment]
Please squash commits, and I'll add the LGTM label

**Comment by jsafrane** [comment]
* To fix review remarks in the other provisioners, I reworked AWS cloud provider to use GiB instead of MiB. AWS EBS internally works with GiB, just the internal Kubernetes cloud provider API from some reason exports MiB.

* Also, I exported `RoundUpSize()` function. Personally, I think the previous version was better. Now GCE and Cinder provisioner pull requests depend on this one (AWS).

**Comment by jsafrane** [comment]
Squashed and ready to merge.

**Comment by jsafrane** [comment]
@saad-ali: it's ready to merge

**Comment by saad-ali** [comment]
ok to merge

**Comment by pmorie** [comment]
@saad-ali thanks a lot for the review

**Comment by pmorie** [comment]
@k8s-oncall could the on-call dev consider this one for manual merge?  Two more PRs depend on it that we would like to get in for OpenShift 3.1.1

**Comment by pmorie** [comment]
Bueller?

**Comment by saad-ali** [comment]
@jlowdermilk who's covering for scheduled on-call

**Comment by jsafrane** [comment]
@saad-ali, thanks for the review! And, what are we waiting for here exactly? Is some bot broken?

**Comment by pmorie** [comment]
@k8s-oncall If we could get a manual merge on this, it would be appreciated.  We have two more PRs to get into master before we can put them responsibly into OpenShift for our 3.1.1 release, and we are technically beyond code cutoff.

**Comment by pmorie** [comment]
@jsafrane the merge queue is blocked because some E2E groups are broken in master.

**Comment by pmorie** [comment]
@jsafrane at least, the queue was blocked yesterday, looks like it's green again now, so hopefully the bot will pick this change up soon.

**Comment by saad-ali** [comment]
The build was unstable/broken much of yesterday, so the merge bot was blocked most of the day. Looks like this PR is in 9th in the queue at the moment.

I'll manually merge to unblock you guys. CC @jlowdermilk who's covering for scheduled on-call.

**Comment by pmorie** [comment]
@saad-ali thanks a lot, appreciated

**Comment by jlowdermilk** [comment]
Sorry I missed your earlier ping(s), was unblocking mergebot. Thanks for merging @saad-ali

**Pull request merged**